### ISSUE TITLE
Do not force initialization of Flask-Babel if it is installed

### ISSUE THREAD

**Issue by sbrun** [Flask-Middleware/flask-security#488]
Hello,

If Flask-Babel is installed on the system, it has to be initialized. If not we get this error:
`ValueError: Flask-Babel or Flask-BabelEx is installed but not initialized`
But it's not really compatible with installation as package in a distribution such as Debian. Flask-Babelex might be required for an application (so it's installed on the system) and not used by a second application using Flask-Security. The second application has "no reason" to initialize Flask-Babelex but it will raise the initialization error and fails.

Can you change your check to test if Flask-Babelex has been imported by the application instead of trying to import it yourself?

**Comment by jwag956** [comment]
Flask-Security as part of initialization, always attempts to import Flask-Babel, if not found, then attempts to import Flask-BabelEx. There is (and never has been) a way to configure Flask-Security to not attempt to use Babel.

I would suggest that an application should and needs to have their own environment using  virtualenv, poetry, conda, etc to manage dependencies.

**Comment by sbrun** [comment]
Trying to import Flask-BabelEx or Flask-Babel is not the issue. The issue is that if Flask-BabelEx is installed but not initialized by the application, it fails.
I haven't this issue with version < 4.0.0.

I understand that developers always prefer that the application uses a specific environment but it's not how the distributions handle and install the applications and the Python modules.

**Comment by jwag956** [comment]
Fair enough - as the release notes state - Flask-Babel recently put out a supported release that includes the important features of Flask-BabelEx (which is no longer maintained and has been archived).
Flask-BabelEx would 'automatically' initialize itself - however Flask-Babel doesn't - thus throwing a very confusing and difficult to debug error message somewhere deep down in a request.

So - prior to 4.0 - Flask-Security just supported Flask-BabelEx which didn't require any initialization - but WAS used. In 4.0 Flask-Babel does require initialization - thus the check.

I could remove the check for Flask-BabelEx since that isn't strictly necessary.

I am interested in what distribution is installing Flask-Security and how they are managing required dependencies?

**Comment by sbrun** [comment]
Yes I think it would be a good thing to remove the check for Flask-BabelEx.

I worked on Kali, a Debian derivative. Flask-Security comes from Debian: the latest packaged version (4.0.0) does not depend on Flask-Babel or Flask-BabelEx.
But the previous version (3.4.2) depended on Flask-BabelEx. So when I upgraded the system, Flask-security has been upgraded but Flask-BabelEx remains installed on the machine. And I got the failure...

**Comment by rhertzog** [comment]
(I'm also working on Kali with @sbrun )

> Flask-BabelEx would 'automatically' initialize itself - however Flask-Babel doesn't - thus throwing a very confusing and difficult to debug error message somewhere deep down in a request.
> So - prior to 4.0 - Flask-Security just supported Flask-BabelEx which didn't require any initialization - but WAS used. In 4.0 Flask-Babel does require initialization - thus the check.

Given the restriction that you mention, it seems weird to have flask security load Flask-Babel just because it's available... you're sure to break the application if the application is not using it. (That said this doesn't seem to match what we have been experiencing, we only had the failure with Flask-BabelEx installed but not used... not the opposite, not sure why)

Wouldn't it be better to run the check only if the module has been loaded by the application?
e.g with a check like `if 'flask_babel` in sys.modules`

In other words, I would suggest to not use flask_babel at all if you are in the situation where flask_babel has not been loaded (and thus not initialized).

**Comment by jwag956** [comment]
Thanks for the additional info. This isn't as clearcut as it seems.
There are 2 things at work here:

First - Prior to 4.0 Flask-Security required Flask-BabelEx, and Flask-BabelEx didn't actually require the application to explicitly initialize it (it 'self' initialized).

Last year, Flask-Babel finally incorporated the key changes from Flask-BabelEx and now Flask-BabelEx is no longer supported/maintained.
Flask-Babel DOES require an explicit initialization.
In 4.0 Flask-Security first looks for Flask-Babel, if that isn't present then looks for Flask-BabelEx, and if that isn't present disables translations services. So it doesn't require/depend on any of them - just uses them if available.

We can't just look for Flask-Babel in sys.modules since Flask-Security ITSELF imports Flask-Babel (attempts to anyway) as part of initial import.

Having Flask-Security look to see if Flask-Babel has been initialized before using it means we are forcing ordering in the application which may cause other backward compatibility and circular dependencies.
Removing the check  won't change the fact that Flask-Security will use it if it finds it - just that Flask-BabelEx will 'just work' and Flask-Babel will fail.  Removing the check would be backwards compatible - I would leave it in for Flask-Babel - however I am not certain that actually addresses your underlying issues.

I will note that Flask-Admin has the same basic algorithm.

**Comment by rhertzog** [comment]
> Having Flask-Security look to see if Flask-Babel has been initialized before using it means we are forcing ordering in the application which may cause other backward compatibility and circular dependencies.

If you do that check at import time, yes you risk having that kind of issues, but if you do it at runtime as part of some "proxy" ot the flask_babel Domain class, then you might avoid the issue (at the cost of redoing the check on a regular basis until we figure out it has been initialized).

But I agree that neither approach is fully satisfying and that this is a service that flask_babel itself should be providing.
